### PR TITLE
TFDataset support nested structure of tensors

### PR DESCRIPTION
**Pull request by yangw1234**

### CODE REVIEW

**Comment by yangw1234** [comment]
Hi @jason-dai, is this api more clear?

For training, we enforce the element of rdd to be a tuple two -- **(features, labels)**. The features or labels can be a single tensor or a dict of tensors.

**Use Case 1**, single feature, single label, should cover most scenarios
```
dataset = TFDataset.from_rdd(training_rdd,
                                 feature = (tf.float32, (28, 28, 1),
                                 labels = (tf.int32, ()),
                                 batch_size=280
                                 )
```
The element of corresponding rdd should be **a tuple of two tensors**, the first is feature, the second is label.


**Use Case 2**, multiple features, single label
```
dataset = TFDataset.from_rdd(training_rdd,
                                 feature = {
                                            "input1": (tf.float32, (28, 28, 1))
                                            "input2": (tf.float32, (28, 28, 1))
                                                 },
                                 labels = (tf.int32, ()),
                                 batch_size=280
                                 )
```
The element of corresponding rdd should be **a tuple of two, the first component is a dict representing feature, the second is a tensor representing label**.


**Use Case 3**, multiple features, multiple labels
```
dataset = TFDataset.from_rdd(training_rdd,
                                 feature = {
                                            "input1": (tf.float32, (28, 28, 1))
                                            "input2": (tf.float32, (28, 28, 1))
                                                 },
                                 labels = {
                                            "label1": (tf.int32, ())
                                            "label2": (tf.int32, ())
                                                 },
                                 batch_size=280
                                 )
```
The element of corresponding rdd should be **a tuple of two, the first component is a dict representing feature, the second is also a dict representing label**.

**Use Case 4**, no label for prediction
```
dataset = TFDataset.from_rdd(training_rdd,
                                 feature = (tf.float32, (28, 28, 1)) or {
                                            "input1": (tf.float32, (28, 28, 1))
                                            "input2": (tf.float32, (28, 28, 1))
                                                 },
                                 batch_size=280
                                 )
```
The element of corresponding rdd should be **a single tensor or a single dict**.

**Comment by jason-dai** [comment]
How do you plan to map the previous API to the new one? @yangw1234

**Comment by yangw1234** [comment]
@jason-dai I'm thinking letting the two versions of API coexist. If a user choose to use the old API, he/she should provide a model_fn with following signature:
```python
def model_fn(tensors, mode):
     ...
     return EstimatorSpec(mode, predictions, loss, optimizer, eval_metric_ops)
```
And log a warning stating this API is deprecated.


If the new API
```python
def model_fn(features, labels, mode):
    ...
    return EstimatorSpec(mode, predictions, loss, optimizer, eval_metric_ops)
```

**Comment by yangw1234** [comment]
jenkins pass http://172.168.2.101:8080/job/ZOO-PR-Validation-Spark-2.4/217/

File: pyzoo/zoo/pipeline/api/net.py [hunk 1]
```diff
@@ -684,47 +684,226 @@ def __init__(self, rdd, names, shapes, types, batch_size,
...
+
+    @property
+    def feature_tensors(self):
+
+        if self._tensors is None:
+            tensors = self._create_placeholders()
+            self._tensors = tensors
+
+        if not isinstance(self._tensors, tuple):
+            raise ValueError("To use feature_tensors, " +
```
Comment (jason-dai) [inline pyzoo/zoo/pipeline/api/net.py]: Ask the users to use the method `from_rdd(training_rdd, features = ..., labels = ..., ...)` here

File: pyzoo/zoo/pipeline/api/net.py [hunk 2]
```diff
@@ -684,47 +684,226 @@ def __init__(self, rdd, names, shapes, types, batch_size,
...
+        if self._tensors is None:
+            tensors = self._create_placeholders()
+            self._tensors = tensors
 
-        self.rdd = rdd.map(lambda arr: arr[:len(names)])
-        self.input_names = names
-        for i in range(len(self.tensors)):
-            tf.add_to_collection(self.tensors[i].name, self)
+        if not isinstance(self._tensors, tuple):
+            raise ValueError("To use label_tensors, the element in " +
```
Comment (jason-dai) [inline pyzoo/zoo/pipeline/api/net.py]: Ask the users to use the method `from_rdd(training_rdd, features, labels, ...)` here

File: pyzoo/zoo/pipeline/api/net.py [hunk 3]
```diff
@@ -684,47 +684,226 @@ def __init__(self, rdd, names, shapes, types, batch_size,
...
+                tensor_structure.append(TensorMeta(types[i], name=names[i], shape=shapes[i]))
+        else:
+            tensor_structure = [TensorMeta(dtype=tf.float32), TensorMeta(dtype=tf.float32)]
+
+        return TFDataset(rdd, tensor_structure,
                          batch_size, batch_per_thread,
                          hard_code_batch_size, val_rdd)
 
+    @staticmethod
+    def from_tuple_rdd(rdd, features, labels=None, batch_size=-1, batch_per_thread=-1,
```
Comment (jason-dai) [inline pyzoo/zoo/pipeline/api/net.py]: `from_tuple_rdd` is a confusing name; we should combine the above two methods into one `from_rdd` method

File: pyzoo/zoo/pipeline/api/net.py [hunk 4]
```diff
@@ -684,47 +684,226 @@ def __init__(self, rdd, names, shapes, types, batch_size,
...
+        else:
+            label_structure = None
+
+        tensor_structure = (feature_structure, label_structure)
+        TFDataset(rdd, tensor_structure,
+                  batch_size, batch_per_thread,
+                  hard_code_batch_size, val_rdd)
+
+    @staticmethod
+    def from_tensors(tensors, batch_size=-1, batch_per_thread=-1,
```
Comment (jason-dai) [inline pyzoo/zoo/pipeline/api/net.py]: why do we need this method?

**Pull request merged**